### ISSUE TITLE
provider/aws: Updating aws_appautoscaling_policy does not update the attached aws_cloudwatch_metric_alarm

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#486]
_This issue was originally opened by @xsellier as hashicorp/terraform#11414. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi,

Since this bug appears in a very complex deployment configuration, I'm creating a github repository to reproduce this bug thanks to a git clone. (see https://github.com/xsellier/terraform-bug-11414)

### Terraform Version
Terraform v0.8.3-dev (4fd10356221a6880c62c86d0d2465f692b8b61bf)

### Affected Resource(s)

- aws_cloudwatch_metric_alarm
- aws_appautoscaling_policy

### Terraform Configuration Files
https://github.com/xsellier/terraform-bug-11414/blob/master/main.tf

### Debug Output
- First `terraform apply`
https://github.com/xsellier/terraform-bug-11414/blob/master/first_output.log

- Second `terraform apply`
https://github.com/xsellier/terraform-bug-11414/blob/master/second_output.log

### Expected Behavior
- The autoscaling group is updated

### Actual Behavior
- The autoscaling group is updated
- The cloudwatch alarm action is updated (because the arn has changed)

### Steps to Reproduce
1. Download https://github.com/xsellier/terraform-bug-11414/archive/master.zip
2. Extract `master.zip`
3. Run `./test.sh`

**Comment by bflad** [comment]
Hi folks 👋 Sorry for the unexpected behavior here. It turns out there was a very subtle bug that prevented proper recreation of Application AutoScaling Policies when only the `resource_id` argument changed. We also now properly ignore `ObjectNotFoundException` errors during Terraform resource destroy. A full writeup of these changes can be found in #7982. These fixes will be released in version 2.3.0 of the Terraform AWS Provider in the next day or two.

Even without upgrading your Terraform AWS Provider to the newer version, you _may_ be able to workaround the original issue with a simple configuration update. Using direct references to the relevant `aws_appautoscaling_target` resource for the `resource_id`, `scalable_dimension`, and `service_namespace` arguments in the `aws_appautoscaling_policy` resource configuration should be enough in many cases to ensure Terraform has proper ordering information to handle recreations of the Application AutoScaling Policy resource when the underlying Application AutoScaling Target or resource associated with the target changes:

```hcl
resource "aws_appautoscaling_target" "example" {
  # ... other configuration ...
}

resource "aws_appautoscaling_policy" "example" {
 # ... other configuration ...

  resource_id        = "${aws_appautoscaling_target.example.resource_id}"
  scalable_dimension = "${aws_appautoscaling_target.example.scalable_dimension}"
  service_namespace  = "${aws_appautoscaling_target.example.service_namespace}"
}
```

If you are still having trouble after upgrading to version 2.3.0 of the Terraform AWS Provider (when its released) and with a configuration looking similar to the above, please create a new GitHub issue with the relevant details from the issue template and we can further triage. Thanks!

**Comment by bflad** [comment]
This has been released in [version 2.3.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#230-march-21-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
